### PR TITLE
Feat/advanced templates

### PR DESCRIPTION
**Pull request by anngvu**
- #338. Why:
  - This is **an under-the-hood change that allows cleaner and more consistent reusable building blocks**. I think I tried to do this within what schematic provided, but it didn't work as well as I'd like and felt like a mistake. This is conceptually nicer. 
  - **Easier to compare templates and see how they are different/specialized**. See more meaningful relationship between templates through the inheritance hiearchy.
  - One **possible cons** is that because it is easier to make templates now, we'll have **over-proliferation** of templates (some more specific than we need, so need to keep that in mind.)
- #251 / #167. Why:
  - Yes, just don't have these in non-applicable templates (like for protocols). Looks more obvious what's happening now.
- #339. Why:
  - While **technically not able to be used by schematic yet _in the future_** it might actually be useful for improved UI and validation. 
  - **Outside of schematic, *we* can possibly use these constraints to make recommendations on Synapse/in custom validation workflow (nfportalutils)**. If we only know about assay, the default is to suggest the raw data template for that assay. If we know both assay and datatype (ultimately, datatype is more important), we can do a better job of directing contributors to the most appropriate template 🥺 
  - **Additional clarity**, helps to realize whether or not all expected assays/datatypes **are actually accounted for** in the available template selection...
- #114 / #168. Why:
  - Technically, we already started doing this for NF-processed data. Having these mappings/alignments is helpful and can improve quality. 
  - Perhaps provide starting example for cross-team data model mapping.
- #138 -> "libraryStrand". Why:
  - This was just overdue, but this change will "improve quality".
- #343. Why:
  - Assurance that everything still works with latest stable version of schematic!

**Update:** 
- Added tests, not sure if Christina has gotten a pre-official-review time as a heads-up for how proposed changes might fit into this, but I'll schedule some time for interactive working session as well where we can experiment with a fork of this.

### CODE REVIEW

File: modules/Assay/Assay.yaml [hunk 1]
```diff
@@ -9,6 +9,8 @@ enums:
         meaning: https://www.thermofisher.com/us/en/home/life-science/cell-analysis/fluorescence-microplate-assays/microplate-assays-cell-viability/alamarblue-assay-cell-viability.html
       2D CellTiter-Glo:
         description: Cell viability assay based on detection of ATP.
+        todo:
+        - This should be subsumed under `cell viability assay`, can add CellTiter-Go as platform.
```
Comment (allaway) [inline modules/Assay/Assay.yaml:13]: Maybe same for 2D incucyte below?

File: modules/Assay/Parameter.yaml [hunk 2]
```diff
@@ -44,111 +17,153 @@ enums:
...
-        description: Fluorescent dye used for the red channel on an array
-        meaning: http://purl.obolibrary.org/obo/FBbi_00000450
-      Cy3:
-        description: Fluorescent dye used for the green channel on an array
-        meaning: http://purl.obolibrary.org/obo/FBbi_00000449
-  GenePerturbationTechnologyEnum:
+      singleEnd:
+        description: A library preparation that results in the creation of a library of 5' ends of DNA.
+        meaning: Sage Bionetworks
+  StrandednessEnum:
```
Comment (allaway) [inline modules/Assay/Parameter.yaml:155]: Do you think it is worth making this identical to HTAN's implementation of this concept? 
eg

term: Library Strand
values: Unstranded, FirstStranded, SecondStranded, NotApplicable

File: modules/Assay/Assay.yaml (hunk 1 above)
Comment (anngvu) [reply modules/Assay/Assay.yaml:13]: Yeah, agree

File: modules/Assay/Parameter.yaml (hunk 2 above)
Comment (anngvu) [reply modules/Assay/Parameter.yaml:155]: Done

**Pull request merged**